specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-08-01-preview/workspaceRP.json

```diff
@@ -3138,13 +3205,25 @@
           "description": "Category of the connection",
           "$ref": "#/definitions/ConnectionCategory"
         },
+        "createdByWorkspaceArmId": {
+          "type": "string"
+        },
         "expiryTime": {
           "format": "date-time",
           "type": "string"
         },
+        "isSharedToAll": {
+          "type": "boolean"
+        },
         "metadata": {
```

Comment:
>metadata [](http://example.com/codeflow?start=9&length=8)

From ARM perspective, this is moving the wrong direction. The empty/unschematized object pattern is not recommended. Generally, this pattern should only be used when needed and the object properties are dynamic, unknown, or user defined. Ref: [https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi032-only-use-additionalproperties-when-the-object-properties-are-dynamic-unknown-or-user-defined](https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi032-only-use-additionalproperties-when-the-object-properties-are-dynamic-unknown-or-user-defined).

The properties of this object can't be supported by core ARM features like Azure Policy and Azure Resource Graph. Customers expect those features to work on your APIs. Consider how you might apply some schematization to the metadata for this object, rather than removing it entirely.

Reply:
Thanks for the feedback. Yes, this metadata property is more like a user defined object, we do have it defined in server side but it's kind of recursive: the metadata will be a Dictionary<string, IValue> where IValue can be a string or another layer of Dictionary<string, IValue>, I don't know if we support recursive schema, let me know what should be the best way to move forward.

Reply:
Recursive schema is no problem, but a schema of string-or-object is a bigger problem. The closest thing to that is the discriminator pattern, which adds some complexity, and doesn't support arbitrary depth of the tree. I guess the bigger question is why such complex metadata is required. This is very difficult for clients to consume. How does a given client know whether a given value is (or should be) present? It's essentially arbitrary JSON, where your clients don't even know how deep the tree goes. Taking a dependency on values in this kind of metadata will be likely to cause customer code breaks. Does your RP also read/write the metadata, or just store it for the client?

Reply:
I see, after discussion with my team we will make this metadata a one layer Dictionary<string, string>

Reply:
The justification to use a dictionary is unclear. Please provide a clear justification or book an ARM modelling office hours slot to explain your scenario

Reply:
This metadata field is a customer-defined field used for the customer to store info they want, to us it's like a JOBject, so that's why we define it as an "object" before, but @mentat9 have some comment above saying we shouldn't do that so we think we can limit this to a one layer dictionary instead, there is no intend to change this metadata field in this version  at all, so if you think the dictionary is not good as well I'm happy to revert it back to an object and discuss what this should be in an office hour. but this property is already in several previouse preview version swaggers I don't think we should block this current version

Reply:
changes related with metadata has been reverted